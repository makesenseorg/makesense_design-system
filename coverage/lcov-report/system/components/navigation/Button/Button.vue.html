<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for system/components/navigation/Button/Button.vue</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../../prettify.css" />
    <link rel="stylesheet" href="../../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../../index.html">All files</a> / <a href="index.html">system/components/navigation/Button</a> Button.vue
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">69.23% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>9/13</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">53.33% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>8/15</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>1/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">69.23% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>9/13</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">&lt;template&gt;
  &lt;component
    @click.capture="handleClick"
    class="ds-button"
    :class="[
      size &amp;&amp; `ds-button-size-${size}`,
      fullwidth &amp;&amp; `ds-button-fullwidth`,
      primary &amp;&amp; `ds-button-primary`,
      secondary &amp;&amp; `ds-button-secondary`,
      danger &amp;&amp; `ds-button-danger`,
      ghost &amp;&amp; `ds-button-ghost`,
      iconOnly &amp;&amp; `ds-button-icon-only`,
      hover &amp;&amp; `ds-button-hover`
    ]"
    v-bind="bindings"
    :is="linkTag"&gt;
    &lt;ds-icon 
      v-if="icon" 
      :name="icon"/&gt;
    &lt;span 
      class="ds-button-text"
      v-if="$slots.default"&gt;
      &lt;slot /&gt;
    &lt;/span&gt;
    &lt;ds-icon 
      v-if="iconRight" 
      :name="iconRight"/&gt;
  &lt;/component&gt;
&lt;/template&gt;
&nbsp;
&lt;script&gt;
/**
 * Used to provide actions or navigation.
 * @version 1.0.0
 */
export default {
  name: 'DsButton',
  props: {
    /**
     * The path of this button. Can be a url or a Vue router path object.
     */
    path: {
      type: [String, Object],
      default() {
        return null
      }
    },
    /**
     * The size used for the text.
     * @options small|base|large
     */
    size: {
      type: String,
      default: null,
<span class="fstat-no" title="function not covered" >      validator: value =&gt; {</span>
<span class="cstat-no" title="statement not covered" >        return value.match(/(small|base|large)/)</span>
      }
    },
    /**
     * The component / tag used for this button
     * @options router-link|a|button
     */
    linkTag: {
      type: String,
      default() {
<span class="branch-0 cbranch-no" title="branch not covered" >        const defaultLink = this.$router ? 'router-link' : 'a'</span>
<span class="branch-0 cbranch-no" title="branch not covered" >        return this.path ? defaultLink : 'button'</span>
      },
      validator: value =&gt; {
        return value.match(/(router-link|a|button)/)
      }
    },
    /**
     * Fill the full width
     */
    fullwidth: {
      type: Boolean,
      default: false
    },
    /**
     * Primary style
     */
    primary: {
      type: Boolean,
      default: false
    },
    /**
     * Secondary style
     */
    secondary: {
      type: Boolean,
      default: false
    },
    /**
     * Danger style
     */
    danger: {
      type: Boolean,
      default: false
    },
    /**
     * Toggle the hover state
     */
    hover: {
      type: Boolean,
      default: false
    },
    /**
     * Make the buttons background transparent
     */
    ghost: {
      type: Boolean,
      default: false
    },
    /**
     * The name of the buttons icon.
     */
    icon: {
      type: String,
      default: null
    },
    /**
     * The name of the buttons right icon.
     */
    iconRight: {
      type: String,
      default: null
    }
  },
  computed: {
    bindings() {
      const bindings = {}
<span class="branch-1 cbranch-no" title="branch not covered" ><span class="missing-if-branch" title="if path not taken" >I</span>      if (this.path &amp;&amp; this.linkTag === 'router-link') {</span>
<span class="cstat-no" title="statement not covered" >        bindings.to = this.path</span>
      }
<span class="branch-1 cbranch-no" title="branch not covered" ><span class="missing-if-branch" title="if path not taken" >I</span>      if (this.path &amp;&amp; this.linkTag === 'a') {</span>
<span class="cstat-no" title="statement not covered" >        bindings.href = this.path</span>
      }
      return bindings
    },
    iconOnly() {
<span class="branch-2 cbranch-no" title="branch not covered" >      return !this.$slots.default &amp;&amp; this.icon &amp;&amp; !this.iconRight</span>
    }
  },
  methods: {
    handleClick(event) {
      /**
       * Click on button.
       * Receives two arguments:
       * event, route object
       *
       * @event click
       */
<span class="cstat-no" title="statement not covered" >      this.$emit('click', event, this.route)</span>
    }
  }
}
&lt;/script&gt;
&nbsp;
&lt;style lang="scss" src="./style.scss"&gt;
&lt;/style&gt;
&nbsp;
&lt;docs src="./demo.md"&gt;&lt;/docs&gt;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Tue Jan 22 2019 22:02:51 GMT+0100 (GMT+01:00)
</div>
</div>
<script src="../../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../../sorter.js"></script>
</body>
</html>
