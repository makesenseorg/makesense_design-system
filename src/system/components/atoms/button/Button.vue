<template>
  <component
    :is="tag"
    :class="[
      `button button--type-${type} button--size-${size}`,
      { 'button--loading': loading }
    ]"
    :role="role"
    tabindex="0"
    :type="inputType"
    :disabled="disabled"
    ref="container"
    :style="{
      minWidth: $slots.loading === undefined ? minWidth + 'px' : 'auto'
    }"
    @click="onClick"
    @keyup.enter="onClick"
  >
    <mks-icon
      v-if="icon && !iconRight"
      :type="icon"
      class="button__icon button__icon--left"
      :size="size === 'round' ? 24 : 20"
    ></mks-icon>
    <!-- @slot Content of the button-->
    <slot v-if="!loading" />
    <!-- @slot Content of the button when `loading` prop is true -->
    <slot v-else name="loading">
      <mks-loading
        type="standalone"
        :inline="true"
        color="#fff"
        :size="20"
      ></mks-loading>
    </slot>
    <mks-icon
      v-if="icon && iconRight"
      :type="icon"
      class="button__icon button__icon--right"
      :size="size === 'round' ? 24 : 20"
    ></mks-icon>
  </component>
</template>
<script>
/**
 * Buttons can be used for any action or inner link.
 * @version 0.3.0
 */
export default {
  name: "MksButton",
  release: "0.3.0",
  props: {
    /**
     * Buttons can be replaced by any of these tags : "div", "a", "input", "button", "span"
     */
    tag: {
      type: String,
      default: "button",
      validator: function(value) {
        return ["div", "a", "input", "button", "span"].indexOf(value) !== -1;
      }
    },
    /**
     * If loading is set to true, the button will display the loading slot or "Processing..."
     */
    loading: {
      type: Boolean,
      default: false
    },
    /**
     * If disabled is set to true, the button will be disabled
     */
    disabled: {
      type: Boolean,
      default: false
    },
    /**
     * Can set the input type, in case of button tag (ex: "submit")
     */
    inputType: {
      type: String,
      required: false,
      default: ""
    },
    /**
     * Color of the button. "primary",
            "secondary",
            "tertiary",
            "neutral",
            "positive",
            "warning",
            "negative"
     */
    type: {
      type: String,
      default: "primary",
      validator: function(value) {
        return (
          [
            "primary",
            "secondary",
            "tertiary",
            "neutral",
            "positive",
            "warning",
            "negative",
            "text",
            "text-active"
          ].indexOf(value) !== -1
        );
      }
    },
    /**
     * Has to be a valid icon type
     */
    icon: {
      type: String
    },
    /**
     * Place the icon on the right side of the text
     */
    iconRight: {
      type: Boolean,
      default: false
    },
    /**
     * Size of the button. "default", "small", "medium" "full", "round"
     */
    size: {
      type: String,
      default: "default",
      validator: function(value) {
        return (
          ["default", "small", "medium", "full", "round", "square"].indexOf(
            value
          ) !== -1
        );
      }
    },
    /**
     * Size of the button. "default", "small", "full", "round"
     */
    role: {
      type: String,
      default: "button"
    }
  },
  data: function() {
    return {
      minWidth: 0,
      observer: null
    };
  },
  mounted() {
    this.updateMinWidth();

    this.observer = new MutationObserver(() => {
      this.updateMinWidth();
    });

    this.observer.observe(this.$refs.container, {
      attributes: true,
      childList: true,
      characterData: true,
      subtree: true
    });
  },
  beforeDestroy: function() {
    // Clean up
    this.observer.disconnect();
  },
  methods: {
    onClick() {
      /**
       * Fires after click.
       *
       * @event click
       */
      this.$emit("click");
    },
    updateMinWidth() {
      this.$nextTick(function() {
        var container = this.$refs.container;
        if (container !== undefined) this.minWidth = container.offsetWidth;
      });
    }
  }
};
</script>

<style lang="scss" scoped>
.button {
  @include cursor-pointer;
  @include border-radius;
  @include button-reset;
  @include padding-button;
  @include text-body;
  display: inline-block;
  font-weight: $font-weight-extrabold; // override text-body
  border-width: $border-width-m;
  border-style: solid;
  text-decoration: none !important;
  transition: background-color 0.2s, border-color 0.2s;
  text-align: center;

  &:focus {
    @include focus($background-color-primary);
  }

  &--size-small {
    @include border-round;
    @include padding-button-small;
  }

  &--size-medium {
    @include border-radius;
    @include padding-button-small;
  }
  &--size-full {
    width: 100%;
    text-align: center;
  }

  &--size-round {
    @include title-small;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    user-select: none;
    border-radius: $border-radius-circle;
    padding-right: 0;
    padding-left: 0;
    width: $space-xxl;
    height: $space-xxl;
    text-align: center;

    .button__icon {
      margin: 0;
    }
  }
  &--size-square {
    @include text-body;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    user-select: none;
    border-radius: $border-radius-m;
    padding-right: 0;
    padding-left: 0;
    width: $space-xl;
    height: $space-xl;
    padding: $space-xxs;
    text-align: center;
  }

  @mixin type-text {
    @include text-small;
    background-color: transparent;
    text-transform: uppercase;
    color: inherit;
    padding: 0;
    margin: 0;
    border: none;
  }

  &--type-text {
    @include type-text;
    font-weight: $font-weight-normal;
  }

  &--type-text-active {
    @include type-text;
    font-weight: $font-weight-black;
  }

  &--type-primary {
    background-color: $color-primary;
    border-color: $color-primary;
    color: $color-primary-inverse;

    &:hover {
      background-color: $color-primary-active;
      border-color: $color-primary-active;
    }
  }

  &--type-secondary {
    background-color: $color-secondary;
    border-color: $color-secondary;
    color: $color-secondary-inverse;

    &:hover {
      // todo : garder la border peut-être? ou rendre l'effet un peu plus légeer
      background-color: $color-secondary-active;
      border-color: $color-secondary-active;
    }

    &:focus {
      @include focus($color-secondary);
    }
  }

  &--type-tertiary {
    background-color: $color-tertiary;
    border-color: $color-tertiary;
    color: $color-tertiary-inverse;

    &:hover {
      // todo : garder la border peut-être? ou rendre l'effet un peu plus légeer
      background-color: $color-tertiary-active;
      border-color: $color-tertiary-active;
    }

    &:focus {
      @include focus($color-tertiary);
    }
  }

  &--type-positive {
    background-color: $color-success;
    border-color: $color-success;
    color: $color-success-inverse;

    &:hover {
      background-color: $color-success-active;
      border-color: $color-success-active;
    }

    &:focus {
      @include focus($color-success);
    }
  }

  &--type-warning {
    background-color: $color-warning;
    border-color: $color-warning;
    color: $color-warning-inverse;

    &:hover {
      background-color: $color-warning-active;
      border-color: $color-warning-active;
    }

    &:focus {
      @include focus($color-warning-active);
    }
  }

  &--type-negative {
    background-color: $color-danger;
    border-color: $color-danger;
    color: $color-danger-inverse;

    &:hover {
      background-color: $color-danger-active;
      border-color: $color-danger-active;
    }

    &:focus {
      @include focus($color-danger-active);
    }
  }

  &--type-neutral {
    color: $color-neutral-20;
    background-color: $background-color-base;
    border-color: $background-color-softest;

    &:hover {
      background-color: $background-color-softest;
    }

    &:focus {
      @include focus($background-color-softest);
    }
  }

  &:disabled,
  &[disabled] {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  &--loading {
    opacity: 0.7;
  }
}

.button__icon {
  &--left {
    margin-right: $space-xs;
  }
  &--right {
    margin-left: $space-xs;
  }
}

button {
  outline: none !important;
}
</style>
<docs>
## Colors
  ```jsx
    <mks-button tag="button">Button default</mks-button>
    <mks-button tag="button" type="secondary">Button secondary</mks-button>
    <mks-button tag="button" type="tertiary">Button tertiary</mks-button>
    <mks-button tag="button" type="neutral">Button neutral</mks-button>
    <br>
    <mks-button type="positive">Button positive</mks-button>
    <mks-button type="warning">Button warning</mks-button>
    <mks-button type="negative">Button negative</mks-button>
  ```

## Sizes
```jsx
    <mks-button size="round" icon="menu"></mks-button>
    <mks-button size="small">Button small</mks-button>
    <mks-button size="medium">Button medium</mks-button>
    <mks-button size="default">Button default</mks-button>
    <mks-button size="full">Button full</mks-button>
  ```

## With icons
```jsx
    <mks-button size="round" icon="bell"></mks-button>
    <br>
    <mks-button icon="menu" size="small">Menu</mks-button>
    <mks-button icon="menu" icon-right size="small">Menu</mks-button>
    <br>
    <mks-button icon="star">Sauvegarder</mks-button>
    <mks-button icon="star" icon-right>Sauvegarder</mks-button>
  ```

  ## States
```jsx
    <mks-button v-bind:disabled="true">Disabled</mks-button>
    <br>
    <mks-button v-bind:loading="true">
        Load 
        <template v-slot:loading>Loading...</template>
    </mks-button>
    <br>
    <mks-button v-bind:loading="true">
        Load 
    </mks-button>
    <br>
    <mks-button v-bind:loading="true" size="small">
        Load 
    </mks-button>
  ```
</docs>
