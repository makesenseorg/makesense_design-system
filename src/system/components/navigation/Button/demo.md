## Button types

Use a primary button to draw the users attention to important actions. Use default buttons for less important actions.

A danger button should be used only for destructive actions.

```
<ds-button>Default</ds-button>
<ds-button primary>Primary</ds-button>
<ds-button secondary>Secondary</ds-button>
<ds-button danger>Danger</ds-button>
```

## Ghost buttons

Use a ghost button for secondary actions.

```
<ds-button ghost>Default</ds-button>
<ds-button ghost primary>Primary</ds-button>
<ds-button ghost secondary>Secondary</ds-button>
<ds-button ghost danger>Danger</ds-button>
```

## Button sizes

Use different sizes to create hierarchy.

```
<ds-button size="small">Small</ds-button>
<ds-button>Base</ds-button>
<ds-button size="large">Large</ds-button>
```

## Fullwidth button

Use a fullwidth button for important actions.

```
<ds-button primary fullwidth>Fullwidth button</ds-button>
```

## Button states

A button can take different states.

```
<ds-button>Default state</ds-button>
<ds-button disabled>Disabled state</ds-button>
<ds-button hover>Hover state</ds-button>
```

## Icon buttons

Add an icon to a button to help the user identify the button's action.

```
<ds-button icon="plus" primary>Click me</ds-button>
<ds-button icon-right="plus">Click me</ds-button>
<ds-button icon="plus" primary></ds-button>
<ds-button icon="plus" ghost></ds-button>
```

## Button as links

Provide a path to the button. You can pass a url string or a Vue router path object.

```
<ds-button path="/navigation">Click me</ds-button>
<ds-button :path="{ name: 'Navigation' }">Click me</ds-button>
```