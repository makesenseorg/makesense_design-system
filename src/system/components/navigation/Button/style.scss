.ds-button {
  @include reset;
  position: relative;
  width: auto;
  overflow: visible;
  -webkit-font-smoothing: inherit;
  -moz-osx-font-smoothing: inherit;
  -webkit-appearance: none;
  border: 0;
  cursor: pointer;
  user-select: none;
  font-size: $font-size-base;
  line-height: $line-height-base;
  font-family: $font-family-text;
  font-weight: $font-weight-bold;
  letter-spacing: $letter-spacing-large;
  display: inline-flex;
  vertical-align: middle;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  padding: 0 $space-small;
  height: $input-height;
  border-radius: $border-radius-base;
  box-shadow: $box-shadow-small-inset, $box-shadow-x-small;
  border: $input-border-size solid transparent;
  transition: color $duration-short $ease-out,
    background-color $duration-short $ease-out;

  &:before {
    position: absolute;
    content: '';
    top: $space-xxx-small;
    left: $space-xxx-small;
    right: $space-xxx-small;
    bottom: $space-xxx-small;
    border-radius: $border-radius-base;
    box-shadow: $box-shadow-inset;
    opacity: 0;
    visiblity: hidden;
    pointer-events: none;
  }

  &:after {
    position: absolute;
    content: '';
    top: $space-xxx-small;
    left: $space-xxx-small;
    right: $space-xxx-small;
    bottom: $space-xxx-small;
    border-radius: $border-radius-base;
    border: 1px dotted currentColor;
    opacity: 0;
    visiblity: hidden;
    pointer-events: none;
    transition: all $duration-short $ease-out;
  }

  &:focus {
    outline: none;
    &:after {
      opacity: 0.8;
      visibility: visible;
    }
  }

  &:active {
    &:before {
      opacity: 0.6;
    }
  }

  &::-moz-focus-inner {
    border: 0;
    padding: 0;
  }

  &:disabled {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  // Default colors
  color: $text-color-base;
  background-color: $background-color-softer;

  &:hover,
  &.ds-button-hover {
    color: $text-color-base;
    background-color: $background-color-softer-active;
  }
}

.ds-button-fullwidth {
  width: 100%;
}

.ds-button-primary {
  color: $text-color-primary-inverse;
  background-color: $background-color-primary;

  &:hover,
  &.ds-button-hover {
    color: $text-color-primary-inverse;
    background-color: $background-color-primary-active;
  }

  &:active {
    &:before {
      opacity: 1;
    }
  }
}

.ds-button-secondary {
  color: $text-color-secondary-inverse;
  background-color: $background-color-secondary;

  &:hover,
  &.ds-button-hover {
    color: $text-color-secondary-inverse;
    background-color: $background-color-secondary-active;
  }

  &:active {
    &:before {
      opacity: 1;
    }
  }
}

.ds-button-danger {
  color: $text-color-danger-inverse;
  background-color: $background-color-danger;

  &:hover,
  &.ds-button-hover {
    color: $text-color-danger-inverse;
    background-color: $background-color-danger-active;
  }

  &:active {
    &:before {
      opacity: 1;
    }
  }
}

.ds-button-ghost {
  color: $text-color-base;
  background-color: transparent;
  box-shadow: none;

  &:focus {
    box-shadow: none;
  }

  &:hover,
  &.ds-hover {
    color: $text-color-base;
    background-color: $background-color-soft;
  }

  &:active {
    &:before {
      opacity: 0.6;
    }
  }

  &.ds-button-primary {
    color: $text-color-primary;
  }

  &.ds-button-secondary {
    color: $text-color-secondary;
  }

  &.ds-button-danger {
    color: $text-color-danger;
  }
}

.ds-button-size-small {
  font-size: $font-size-small;
  padding: 0 $space-x-small;
  height: $input-height-small;
}

.ds-button-size-large {
  font-size: $font-size-large;
  padding: 0 $space-base;
  height: $input-height-large;
}

.ds-button-icon-only {
  width: $input-height;
  padding: 0;
  border-radius: $border-radius-rounded;

  &:before, &:after {
    border-radius: $border-radius-rounded;
  }

  &.ds-button-size-small {
    width: $input-height-small;
  }

  &.ds-button-size-large {
    width: $input-height-large;
  }
}

.ds-button-text {
  line-height: inherit;
  display: inline-block;
  white-space: nowrap;
  margin: 0 $font-space-small;

  &:first-child {
    margin-left: 0;
  }

  &:last-child {
    margin-right: 0;
  }
}