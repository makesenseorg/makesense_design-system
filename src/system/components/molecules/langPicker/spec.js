import { shallowMount, createLocalVue } from "@vue/test-utils";
import Comp from "./LangPicker.vue";

import Button from "../../atoms/button/Button.vue";

const localVue = createLocalVue();
localVue.component("MksButton", Button);

describe("langPicker.vue", () => {
  it("matches snapshot", () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        langs: ["French", "English"],
        active: "French",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("emits select lang event", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        langs: ["French", "English"],
        active: "French",
      },
    });

    const button = wrapper.find(".app-lang__button--active");
    expect(button.exists()).toBe(true);

    button.vm.$emit("click");
    await wrapper.vm.$nextTick();

    expect(wrapper.emitted("select")).toBeTruthy();
  });

  it("displays active lang modifier", () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        langs: ["French", "English"],
        active: "French",
      },
    });

    const button = wrapper.find(".app-lang__button--active");
    expect(button.exists()).toBe(true);
    expect(button.text()).toBe("French");
  });
});
