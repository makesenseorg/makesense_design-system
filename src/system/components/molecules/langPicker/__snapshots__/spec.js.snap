// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`langPicker.vue matches snapshot 1`] = `
<ul
  aria-label="a"
  class="app-lang"
  role="listbox"
>
  <li
    class="app-lang__item"
  >
    <mks-button-stub
      aria-selected="true"
      class="app-lang__button app-lang__button--active"
      inputtype=""
      role="option"
      size="default"
      tag="button"
      type="text-active"
    >
      French
    </mks-button-stub>
  </li>
  <li
    class="app-lang__item"
  >
    <mks-button-stub
      class="app-lang__button"
      inputtype=""
      role="option"
      size="default"
      tag="button"
      type="text"
    >
      English
    </mks-button-stub>
  </li>
</ul>
`;
